src/main/java/io/vertx/core/http/impl/pool/Pool.java

```diff
@@ -0,0 +1,315 @@
+    this.maxWeight = maxWeight;
+    this.connector = connector;
+    this.queueMaxSize = queueMaxSize;
+    this.poolClosed = poolClosed;
+    this.all = new HashSet<>();
+    this.available = new ArrayDeque<>();
+    this.connectionAdded = connectionAdded;
+    this.connectionRemoved = connectionRemoved;
+  }
+
+  public synchronized boolean getConnection(Waiter<C> waiter) {
+    if (closed) {
+      return false;
+    }
+    // Enqueue
+    if (capacity > 0 || weight < maxWeight || (queueMaxSize < 0 || waiters.size() < queueMaxSize)) {
+      waiters.add(waiter);
+      checkPending();
+    } else {
+      waiter.handleFailure(waiter.context, new ConnectionPoolTooBusyException("Connection pool reached max wait queue size of " + queueMaxSize));
+    }
+    return true;
+  }
+
+  private void checkPending() {
+    while (true) {
+      Waiter<C> waiter = waiters.peek();
+      if (waiter == null) {
+        break;
+      }
+      if (capacity > 0) {
+        capacity--;
+        Holder<C> conn = available.peek();
+        if (--conn.capacity == 0) {
+          available.poll();
+        }
+        waiters.poll();
+        ContextImpl ctx = conn.context;
+        ctx.nettyEventLoop().execute(() -> {
+          if (connector.isValid(conn.connection)) {
+            boolean handled = deliverToWaiter(conn, waiter);
+            if (!handled) {
+              synchronized (Pool.this) {
+                recycleConnection(conn, 1,false);
+                checkPending();
+              }
+            }
+          } else {
+            synchronized (Pool.this) {
+              waiters.add(waiter);
+              closed(conn);
+            }
+          }
+        });
+      } else if (weight < maxWeight) {
+        waiters.poll();
+        weight += createConnection(waiter);
+      } else {
+        break;
+      }
+    }
+  }
+
+  private long createConnection(Waiter<C> waiter) {
+    Holder<C> holder  = new Holder<>();
+    ConnectionListener<C> listener = new ConnectionListener<C>() {
+      @Override
+      public void onConnectSuccess(C conn, long concurrency, Channel channel, ContextImpl context, long initialWeight, long actualWeight) {
+        waiter.initConnection(context, conn);
+        boolean usable;
+        synchronized (Pool.this) {
+          usable = initConnection(waiter, holder, context, concurrency, conn, channel, initialWeight, actualWeight);
+        }
+        if (usable) {
+          boolean consumed = deliverToWaiter(holder, waiter);
+          synchronized (Pool.this) {
+            if (!consumed) {
+              synchronized (this) {
+                recycleConnection(holder, 1,false);
+              }
+            }
+            checkPending();
+          }
+        } else {
+          synchronized (Pool.this) {
+            checkPending();
+          }
+        }
+      }
+      @Override
+      public void onConnectFailure(ContextImpl context, Throwable err, long weight) {
```

Comment:
So if I understand correctly, when a `waiter` is picked up, and no connection is available, but `maxWeight` is not reached, we create a new connection. And if connection fails, the `waiter` handles the failure immediately. Correct?

Why not leaving the waiter in the list and pick it up only when a new connection is ready? Of course the waiter would timeout at some point. I'm not saying this strategy is better, just curious why you picked the other strategy.

Reply:
because it is simpler this way